Async Runner with Message Channel

flashdesignory opened the pull request:
As discussed in our sync, here is a combination of the [AsyncRunner](https://github.com/WebKit/Speedometer/pull/442) and [Message Channel](https://github.com/WebKit/Speedometer/pull/457) pr.

Relevant code snippet (udpated):

```
class AsyncRAFTestInvoker extends BaseRAFTestInvoker {
    static mc = new MessageChannel();
    _scheduleCallbacks(resolve) {
        let gotTimer = false;
        let gotMessage = false;
        let gotPromise = false;

        const tryTriggerAsyncCallback = () => {
            if (!gotTimer || !gotMessage || !gotPromise)
                return;

            this._asyncCallback();
            setTimeout(async () => {
                await this._reportCallback();
                resolve();
            }, 0);
        };

        requestAnimationFrame(async () => {
            await this._syncCallback();
            gotPromise = true;
            tryTriggerAsyncCallback();
        });

        requestAnimationFrame(() => {
            setTimeout(() => {
                gotTimer = true;
                tryTriggerAsyncCallback();
            });

            AsyncRAFTestInvoker.mc.port1.addEventListener(
                "message",
                function () {
                    gotMessage = true;
                    tryTriggerAsyncCallback();
                },
                { once: true }
            );
            AsyncRAFTestInvoker.mc.port1.start();
            AsyncRAFTestInvoker.mc.port2.postMessage("speedometer");
        });
    }
}
```

For testing purposes I kept an explicit type of "async" in the test file, but I opted all default suites into using it. 

Initial testing shows that it should also fix the react specific issue.

Other issue solved by this pr: https://github.com/WebKit/Speedometer/issues/83

flashdesignory commented on resources/test-invoker.mjs:
```diff
@@ -48,8 +67,51 @@ export class RAFTestInvoker extends TestInvoker {
+
+export const ASYNC_TEST_INVOKER_LOOKUP = {
+    __proto__: null,
+    timer: AsyncTimerTestInvoker,
```
If we don't want an AsyncTimerTestInvoker, I could simply assign the same AsyncRAFTestInvoker..

smaug---- commented on resources/test-invoker.mjs:
```diff
@@ -48,8 +67,51 @@ export class RAFTestInvoker extends TestInvoker {
+class AsyncRAFTestInvoker extends BaseRAFTestInvoker {
+    _scheduleCallbacks(resolve) {
+        requestAnimationFrame(async () => {
+            await this._syncCallback();
```
_syncCallback and _asyncCallback should be rename to something else,  since, well, the patch makes _syncCallback rather asynchronous :)

flashdesignory replied:
LOL, we had the same convo in a similar pr... what do you think about:

```
class TestInvoker {
    constructor(captureTest, captureLayout, reportResults) {
        this._captureTest = captureTest;
        this._captureLayout = captureLayout;
        this._reportResults = reportResults;
    }
} 
```

smaug---- commented on resources/test-invoker.mjs:
```diff
@@ -48,8 +67,51 @@ export class RAFTestInvoker extends TestInvoker {
+                tryTriggerAsyncCallback();
+            });
+
+            const mc = new MessageChannel();
```
It might be nice to use the same channel for all the tests, just to avoid extra garbage created by the runner. Not sure where there channel could live. Perhaps a static property on the class?

flashdesignory replied:
are you thinking maybe something like:
```
class AsyncRAFTestInvoker extends BaseRAFTestInvoker {
    static mc = new MessageChannel();
    _scheduleCallbacks(resolve) {
        requestAnimationFrame(async () => {
            await this._syncCallback();

            let gotTimer = false;
            let gotMessage = false;

            const tryTriggerAsyncCallback = () => {
                if (!gotTimer || !gotMessage)
                    return;

                this._asyncCallback();
                setTimeout(async () => {
                    await this._reportCallback();
                    resolve();
                }, 0);
            };

            const handleMessage = () => {
                AsyncRAFTestInvoker.mc.port1.removeEventListener("message", handleMessage);
                gotMessage = true;
                tryTriggerAsyncCallback();
            };

            setTimeout(() => {
                gotTimer = true;
                tryTriggerAsyncCallback();
            });

            AsyncRAFTestInvoker.mc.port1.addEventListener("message", handleMessage);
            AsyncRAFTestInvoker.mc.port1.start();
            AsyncRAFTestInvoker.mc.port2.postMessage("speedometer");
        });
    }
}
```

flashdesignory replied:
@smaug---- and following up here as well in regards to a static MessageChannel

smaug---- commented on resources/test-invoker.mjs:
```diff
@@ -48,8 +67,51 @@ export class RAFTestInvoker extends TestInvoker {
+            setTimeout(() => {
+                gotTimer = true;
+                tryTriggerAsyncCallback();
+            });
```
oh, hmm, does this non-double rAF setup have a bug after all.  If the promise callbacks create new promises which get resolve and those callbacks then trigger setTimeout or postMessage, does anything guarantee that test runners setTimeout/messageport gets handled after those?

flashdesignory replied:
We are stopping and waiting for `await this._syncCallback();`. 
The setTimeout / messageport happens after that resolved, correct?
The reason of having the boolean flags is that we can't guarantee in which order setTimeout / postMessage gets called. At least that's how I understood our conversation 🙂

flashdesignory replied:
I did test it with the current tests as well as with a test that had an artificial promise resolving after a long period of time in each step.

smaug---- replied:
But await this._syncCallback(); might return in the same task, if one just does effectively
return new Promise(r => r());

In my example using MessageChannel there were still two rAFs.

I thought we were thinking to use the same runner for async tests and non-async, and I'm worried that not having two rafs would broke non-async tests (meaning that some time wouldn't be measured correctly).  But I could be missing something (I think I did when I said on Slack couple of days ago that this approach is ok :)).

flashdesignory replied:
@smaug---- following up on the above snippet. Do you think this would satisfy async steps with two rafs?

smaug---- commented on resources/tests.mjs:
```diff
@@ -125,6 +125,7 @@ Suites.push({
     name: "TodoMVC-JavaScript-ES5",
     url: "resources/todomvc/vanilla-examples/javascript-es5/dist/index.html",
     tags: ["todomvc"],
+    type: "async",
```
So in practice all the tests are now async?

flashdesignory replied:
This can be reverted. For testing I marked all default tests async. 
How do we decide which tests should be async? I know React / Next.js had issues with not capturing all work. Those could be candidates to mark async?

flashdesignory replied:
I'll leave it unchanged (opting every single workload into async steps), until I get more feedback from other folks.

smaug---- commented on resources/tests.mjs:
```diff
@@ -857,11 +869,12 @@ Suites.push({
     },
     tests: [
-        new BenchmarkTestStep("NavigateToUS", (page) => {
+        new BenchmarkTestStep("NavigateToUS", async (page) => {
```
Why is async callback needed here? The function doesn't seem to do anything async, no?

flashdesignory replied:
removed!

rniwa commented on resources/test-invoker.mjs:
```diff
@@ -24,7 +24,24 @@ export class TimerTestInvoker extends TestInvoker {
     }
 }
 
-export class RAFTestInvoker extends TestInvoker {
```
I'd prefer removing the TimerTestInvoker entirely since we've moved to RAF-based measurement in 3 and we haven't seen any issue with it.

flashdesignory replied:
removed!

rniwa commented:
Can we split the change to remove `measurementMethod` into a separate PR? We should also update params.mjs and developer-mode.mjs.

flashdesignory commented:
> Can we split the change to remove `measurementMethod` into a separate PR? We should also update params.mjs and developer-mode.mjs.

ok, reverted for now. I agree a cleanup pr to remove this would be better.

bgrins commented:
@smaug---- I pushed this to https://feature-async-runner-message-channel--speedometer-preview.netlify.app/ for anyone who wants to test.

rniwa commented on resources/shared/test-runner.mjs:
```diff
@@ -13,25 +13,52 @@ export class TestRunner {
+        return `${this.#suite.name}.${this.#test.name}-sync-end`;
+    }
+
+    get asyncStartLabel() {
```
Do we really need to add all these getters now that we're only overriding _runSyncStep?

flashdesignory replied:
I can revert this.

flashdesignory replied:
done

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -24,7 +24,24 @@ export class TimerTestInvoker extends TestInvoker {
 }
 
-export class RAFTestInvoker extends TestInvoker {
+class AsyncTimerTestInvoker extends TestInvoker {
```
Can we instead just delete `TimerTestInvoker` as we discussed in a separate PR?

rniwa replied:
We probably want to land the removal PR first.

flashdesignory replied:
thanks for opening https://github.com/WebKit/Speedometer/pull/466

rniwa commented on resources/suite-runner.mjs:
```diff
@@ -75,7 +75,8 @@ export class SuiteRunner {
                 await this.#client.willRunTest(this.#suite, test);
 
-            const testRunner = new TestRunner(this.#frame, this.#page, this.#params, this.#suite, test, this._recordTestResults);
+            const testRunnerClass = TEST_RUNNER_LOOKUP[this.#suite.type ?? "default"];
```
Instead of changing the test runner based on the suite type for some suites, can we initially add a toggle to developer menu so that we can try out async runner?

camillobruni replied:
To my understanding the agreement in the meeting was to unconditionally enable the new runner for the preact workloads and add the global developer-mode menu in a separate PR.

flashdesignory replied:
correct - below is the list on enabled suites (for now):

-  TodoMVC-React-Complex-DOM,
-  TodoMVC-React-Redux,
-  TodoMVC-Preact-Complex-DOM, (technically not react based, but it's similar)
-  NewsSite-Next,
-  React-Stockcharts-SVG

rniwa replied:
If we're doing that, we need before/after numbers across browsers.

flashdesignory replied:
I'll run crossbench for the ones on the list and post results here.

flashdesignory replied:
Here are metrics:
![Screenshot 2024-12-18 at 5 54 40 PM](https://github.com/user-attachments/assets/daf49729-dac6-4bbc-b4c3-b83a76c82bae)

rniwa replied:
So Safari's total time goes up while the total time for Chrome & Firefox both go down? With that kind of magnitude of a difference, we need a more thorough investigation before we can approve of the change. I suggest we land this method initially as an optional developer-menu only feature instead.

smaug---- reviewed (approved):
Looks reasonable

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -48,8 +67,58 @@ export class RAFTestInvoker extends TestInvoker {
 }
 
+class AsyncRAFTestInvoker extends BaseRAFTestInvoker {
+    static mc = new MessageChannel();
```
Why does this need to be a static variable?

rniwa replied:
Oh I see. Maybe but can't we use an instance variable in that case? We don't re-create TestInvoker for each suite / step, right?

flashdesignory replied:
each test creates a new instance:

https://github.com/WebKit/Speedometer/blob/main/resources/shared/test-runner.mjs#L74-L75

flashdesignory commented:
@rniwa - I believe we addressed all the comments. Anything else missing that I can help with?

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -31,7 +33,56 @@ export class RAFTestInvoker extends TestInvoker {
+        };
+
+        requestAnimationFrame(async () => {
+            await this._syncCallback();
```
It is a bit suspicious to alway trigger the workload in rAF in this asynchronous measurement method.
A real web apps don't necessarily always wait for rAF to trigger any work so for the maximal realism, it might be better if we just triggered sync callback at a pseudo-random time relative to rAF.

rniwa replied:
ok, we can keep the double rAF for now but I think it's worth exploring if we can execute at rAF un-aligned time.

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -31,7 +33,56 @@ export class RAFTestInvoker extends TestInvoker {
+
+        requestAnimationFrame(() => {
+            setTimeout(() => {
+                gotTimer = true;
```
So I think we need to wait for any promises scheduled in setTimeout get resolved so I think we need an await here.

rniwa replied:
What I mean is to do this:
```
await Promise.resolve();
```
This will ensure that any promises scheduled during `setTimeout` will get resolved before we set `gotTimer` to `true`.

flashdesignory replied:
@rniwa - added

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -31,7 +33,56 @@ export class RAFTestInvoker extends TestInvoker {
+                "message",
+                function () {
+                    gotMessage = true;
+                    tryTriggerAsyncCallback();
```
Ditto here. I think we need to wait for any promises that got scheduled as a result of postMessage to get processed before we finish the measurement.

flashdesignory replied:
added

rniwa commented on resources/shared/params.mjs:
```diff
@@ -50,12 +52,14 @@ export class Params {
         this.shuffleSeed = this._parseShuffleSeed(searchParams);
 
         const unused = Array.from(searchParams.keys());
+        console.log("unused", unused);
```
This is logging intentional??

flashdesignory replied:
removed

rniwa commented on resources/tests.mjs:
```diff
@@ -343,6 +343,7 @@ Suites.push({
     name: "TodoMVC-React-Complex-DOM",
     url: "resources/todomvc/architecture-examples/react-complex/dist/index.html#/home",
     tags: ["todomvc", "complex", "complex-default"],
+    type: "async",
```
Wouldn't this mean all these workloads will use async runner now?

rniwa commented on resources/shared/test-invoker.mjs:
```diff
@@ -31,7 +33,58 @@ export class RAFTestInvoker extends TestInvoker {
     raf: RAFTestInvoker,
 };
+
+export const ASYNC_TEST_INVOKER_LOOKUP = {
```
Do we really need a separate lookup table? Now that timer-based measurement is gone,
why don't we just have a single lookup table with "raf" and "async" as measurement methods.

rniwa commented on resources/shared/test-runner.mjs:
```diff
@@ -41,7 +52,7 @@ export class TestRunner {
             performance.mark(syncStartLabel);
             const syncStartTime = performance.now();
-            this.#test.run(this.#page);
+            await this._runSyncStep(this.test, this.page);
```
This seems to have a side effect of always pushing the measurement end until when a promise resolves.
I don't think we want to do that for rAF based measurement method (since that can affect the score)
so maybe we should check the return value of this function and await conditionally?

rniwa replied:
More concretely, the following code logs: 1, 2, 3:
```
(async function() { console.log(1); await (() => { })(); console.log(3); })(); console.log(2);
```
whereas the following code logs 1, 3, 2:
```
(async function() { console.log(1); (() => { })(); console.log(3); })(); console.log(2);
```

i.e. `await` will always delay the execution of the continuation until the end of the current micro-task. While that's highly desirable for async measurement, I don't think we want to affect rAF measurement.

flashdesignory commented:
cleaned up some stuff from the developer menu integration. Previously i only checked if the suite.type is "async", which always returned false, since we removed the async type from all tests. 

Now I am checking both: suite.type === "async" or "params.useAsyncSteps"

rniwa reviewed (approved):
Great. Thank you for addressing all the review comments!
